src/transformers/models/vits/modeling_vits.py

```diff
@@ -0,0 +1,1413 @@
+            else:
+                global_states = torch.zeros_like(hidden_states)
+
+            acts = fused_add_tanh_sigmoid_multiply(hidden_states, global_states, num_channels_tensor)
+            acts = self.dropout(acts)
+
+            res_skip_acts = self.res_skip_layers[i](acts)
+            if i < self.num_layers - 1:
+                res_acts = res_skip_acts[:, : self.hidden_size, :]
+                inputs = (inputs + res_acts) * padding_mask
+                outputs = outputs + res_skip_acts[:, self.hidden_size :, :]
+            else:
+                outputs = outputs + res_skip_acts
+
+        return outputs * padding_mask
+
+    def remove_weight_norm(self):
+        if self.speaker_embedding_size != 0:
+            torch.nn.utils.remove_weight_norm(self.cond_layer)
+        for layer in self.in_layers:
+            torch.nn.utils.remove_weight_norm(layer)
+        for layer in self.res_skip_layers:
+            torch.nn.utils.remove_weight_norm(layer)
+
+
+class VitsResidualCouplingLayer(nn.Module):
+    def __init__(self, config: VitsConfig, mean_only: bool = False):
+        super().__init__()
+        self.half_channels = config.flow_size // 2
+        self.mean_only = mean_only
+
+        self.conv_pre = nn.Conv1d(self.half_channels, config.hidden_size, 1)
+        self.wavenet = VitsWaveNet(config, num_layers=4)
+        self.conv_post = nn.Conv1d(config.hidden_size, self.half_channels * (2 - mean_only), 1)
+
+    def forward(self, inputs, padding_mask, global_conditioning=None, reverse=False):
+        x0, x1 = torch.split(inputs, [self.half_channels] * 2, dim=1)
+        h = self.conv_pre(x0) * padding_mask
+        h = self.wavenet(h, padding_mask, global_conditioning)
+        stats = self.conv_post(h) * padding_mask
+
+        if self.mean_only:
+            mean = stats
+            log_stddev = torch.zeros_like(mean)
+        else:
+            mean, log_stddev = torch.split(stats, [self.half_channels] * 2, dim=1)
+
+        if not reverse:
+            x1 = mean + x1 * torch.exp(log_stddev) * padding_mask
+            outputs = torch.cat([x0, x1], dim=1)
+            log_determinant = torch.sum(log_stddev, [1, 2])
+            return outputs, log_determinant
+        else:
+            x1 = (x1 - mean) * torch.exp(-log_stddev) * padding_mask
+            outputs = torch.cat([x0, x1], dim=1)
+            return outputs
+
+
+class VitsResidualCouplingBlock(nn.Module):
+    def __init__(self, config: VitsConfig):
+        super().__init__()
+        self.flows = nn.ModuleList()
+        for _ in range(config.prior_encoder_num_flows):
+            self.flows.append(VitsResidualCouplingLayer(config, mean_only=True))
+            self.flows.append(VitsFlip())
+
+    def forward(self, inputs, padding_mask, global_conditioning=None, reverse=False):
+        if not reverse:
+            for flow in self.flows:
+                inputs, _ = flow(inputs, padding_mask, global_conditioning)
```

Comment:
This is for training. I removed the code that calls this logic because it requires external dependencies (monotonic-align package) and we're not supporting training yet anyway. However, to make the checkpoints compatible with when we do add training, all the additional layers are already in the modeling code, and the `if not reverse` branches use those.

Reply:
Okay makes sense, thanks for adding it in a defensive way like this